### PR TITLE
Mobile distro backend (WIP)

### PR DESCRIPTION
**Pull request by spaudanjo**
Contributes to: https://trello.com/c/iSwBNksN/932-extend-new-be-for-mobile-distribution-prototype

First PR which enriches the GraphQL API schema and enables the BE to support new Mobile Distribution related prototype features.

### CODE REVIEW

**Comment by spaudanjo** [comment]
Haudi @pylipp :) 
Even though I marked it still as a draft - please feel free to scan through it already raise concerns (and also see the questions in the TODO comments)

File: back/boxtribute_server/graph_ql/resolvers.py [hunk 1]
```diff
@@ -398,13 +487,90 @@ def resolve_create_qr_code(*_, box_label_identifier=None):
...
+    authorize(permission="distro_event:write")
+    return create_distribution_event(user_id=g.user.id, **creation_input)
+
+
+@mutation.field("createDistributionSpot")
+@convert_kwargs_to_snake_case
+def resolve_create_distribution_spot(*_, creation_input=None):
+    authorize(permission="location:write")
+    return create_distribution_spot(
+        user_id=g.user.id, distribution_spot_input=creation_input
```
Comment (pylipp) [inline back/boxtribute_server/graph_ql/resolvers.py:510]: would be good to pass `**creation_input` here to unpack the arguments for the `create_distribution_spot` function

File: back/boxtribute_server/graph_ql/resolvers.py [hunk 2]
```diff
@@ -398,13 +487,90 @@ def resolve_create_qr_code(*_, box_label_identifier=None):
...
+    )
+
+
+@mutation.field("removePackingListEntryFromDistributionEvent")
+@convert_kwargs_to_snake_case
+def resolve_remove_packing_list_entry_from_distribution_event(
+    *_, packing_list_entry_id
+):
+
+    packing_list_entry = PackingListEntry.get(packing_list_entry_id)
```
Comment (pylipp) [inline back/boxtribute_server/graph_ql/resolvers.py:550]: If the ID did not exist, peewee would already throw an exception here, hence the error handling in the next line is redundant. The peewee exception is automatically handled due to `format_database_errors` in `exceptions.py`, raising a bad-user-input GraphQL error.

File: back/boxtribute_server/graph_ql/resolvers.py [hunk 3]
```diff
@@ -398,13 +487,90 @@ def resolve_create_qr_code(*_, box_label_identifier=None):
...
+        # TODO: Discuss error handling approach
+        # * seems to be the first time we are doing GraphQLError throwings here
+        # * also discuss how to handle this in the frontend
+        # (e.g. should we return specific error codes that the FE then
+        # handles with messages?)
+        raise GraphQLError("Packing list entry not found")
+    distribution_event = (
+        DistributionEvent.select()
+        .where(DistributionEvent.id == packing_list_entry.distribution_event)
+        .get()
```
Comment (pylipp) [inline back/boxtribute_server/graph_ql/resolvers.py:561]: Maybe already add `.join(DistributionSpot)` because of line 565

File: back/boxtribute_server/graph_ql/resolvers.py [hunk 4]
```diff
@@ -398,13 +487,90 @@ def resolve_create_qr_code(*_, box_label_identifier=None):
...
+    )
+    authorize(
+        permission="distro_event:write",
+        base_id=distribution_event.distribution_spot.base.id,
+    )
+    # Completed Events should not be mutable anymore
+    if distribution_event.state == DistributionEventState.Completed:
+        raise GraphQLError("Cannot move items to completed distribution event")
+    delete_packing_list_entry(packing_list_entry_id)
```
Comment (pylipp) [inline back/boxtribute_server/graph_ql/resolvers.py:570]: For consistency with the other crud functions you could put the entire validation step into the `delete_...` function

File: back/boxtribute_server/graph_ql/resolvers.py [hunk 5]
```diff
@@ -532,6 +698,39 @@ def resolve_base_locations(base_obj, _):
...
+    return (
+        Location.select()
+        .where(Location.type == LocationType.DistributionSpot)
+        .where(base_filter_condition("location:read"))
+    )
+
+
+@query.field("distributionSpot")
+def resolve_distributions_spot(obj, _, id):
+    distribution_spot = obj.location if id is None else Location.get_by_id(id)
```
Comment (pylipp) [inline back/boxtribute_server/graph_ql/resolvers.py:714]: the `id` argument will always be given, so you can shorten this to `distribution_spot = Location.get_by_id(id)` and drop the `obj` argument. Same for the distroEvent query below

File: back/boxtribute_server/models/crud.py [hunk 6]
```diff
@@ -65,6 +76,161 @@ def create_box(
...
+    with db.database.atomic():
+        # Completed Events should not be mutable anymore
+        distribution_event = DistributionEvent.get_by_id(distribution_event_id)
+        if distribution_event.state == DistributionEventState.Completed:
+            raise GraphQLError("Cannot move items to completed distribution event")
+
+        box = Box.get(Box.label_identifier == box_label_identifier)
+
+        # TODO: Discuss error handling approach:
+        # Ok to throw GraphQL errors in the crud module?
```
Comment (pylipp) [inline back/boxtribute_server/models/crud.py:95]: You could define a custom exception from the `exceptions` module (maybe re-use existing base exceptions). Then raise it here. ariadne will wrap it into a GraphQL error

File: back/boxtribute_server/graph_ql/resolvers.py [hunk 7]
```diff
@@ -101,6 +124,35 @@ def resolve_tags(*_):
     return Tag.select()
 
 
+@query.field("packingListEntry")
+def resolve_packing_list_entry(*_, id):
+    authorize(permission="packing_list_entrie:read")
```
Comment (pylipp) [inline back/boxtribute_server/graph_ql/resolvers.py:129]: This should be called `packing_list_entry`...maybe needs to be taken care of in our Auth0 script too

**Comment by pylipp** [comment]
> Haudi @pylipp :) Even though I marked it still as a draft - please feel free to scan through it already raise concerns (and also see the questions in the TODO comments)

I browsed the changes and added a couple of comments. Re the crud functions I did not have a detailed look but would wait for tests instead.

Re the `Box.items` field: I wouldnt mind if (in a different branch) you update the GraphQL field to `numberOfItems` and in the peewee model use `number_of_items = IntField(column_name="items")`. There's likely to be some renaming in different files in the BE and the FE as a result.

File: back/boxtribute_server/models/crud.py (hunk 6 above)
Comment (spaudanjo) [reply back/boxtribute_server/models/crud.py:102]: 👍

**Comment by spaudanjo** [comment]
@pylipp Agree for a dedicated PR to rename items to numberOfItems. 
I addressed your other points - feel free to have another look and approve :)

File: back/boxtribute_server/graph_ql/resolvers.py [hunk 8]
```diff
@@ -532,6 +679,49 @@ def resolve_base_locations(base_obj, _):
     return Location.select().where(Location.base == base_obj.id)
 
 
+@query.field("distributionSpots")
+def resolve_distributions_spots(base_obj, _):
+    authorize(permission="location:read")
+    return (
+        Location.select()
```
Comment (pylipp) [inline back/boxtribute_server/graph_ql/resolvers.py:686]: you have to add `.join(Base)` here

File: back/boxtribute_server/models/crud.py [hunk 9]
```diff
@@ -65,6 +80,170 @@ def create_box(
...
+            last_modified_on=now,
+            last_modified_by=user_id,
+        )
+
+
+def create_distribution_event(
+    user_id,
+    distribution_spot_id,
+    name,
+    planned_start_date_time,
```
Comment (pylipp) [inline back/boxtribute_server/models/crud.py:214]: This is not required by the GraphQL schema but not optional here, probably should be `Datetime!` in the schema

**Review by pylipp — CHANGES_REQUESTED** [review]
What do you think about moving the content from the `crud` module into 

```
boxtribute_server/
    mobile_distribution/
        __init__.py
        spot.py
        event.py
        packing_list.py
```

File: back/boxtribute_server/graph_ql/inputs.graphql [hunk 10]
```diff
@@ -89,3 +103,10 @@ input ShipmentDetailUpdateInput {
   targetProductId: Int!
   targetLocationId: Int!
 }
+
+input PacklistEntryInput {
```
Comment (pylipp) [inline back/boxtribute_server/graph_ql/inputs.graphql:107]: Can this be named `PackingListEntryInput` for consistency?

File: back/boxtribute_server/models/crud.py [hunk 11]
```diff
@@ -65,6 +80,170 @@ def create_box(
...
+        )
+
+    existing_packing_list_entry = PackingListEntry.get_or_none(
+        PackingListEntry.distribution_event == distribution_event_id,
+        PackingListEntry.product == product_id,
+        PackingListEntry.size == size_id,
+    )
+    if existing_packing_list_entry is not None:
+        if number_of_items > 0:
+            existing_packing_list_entry.number_of_items = number_of_items
```
Comment (pylipp) [inline back/boxtribute_server/models/crud.py:187]: Is it intended to replace the number of items, or rather to increment it?

File: back/boxtribute_server/models/crud.py [hunk 12]
```diff
@@ -65,6 +80,170 @@ def create_box(
...
+        unboxed_items_collection.save()
+        box.save()
+        return unboxed_items_collection
+
+
+def move_box_to_distribution_event(box_label_identifier, distribution_event_id):
+    """Move a box to a distribution event."""
+    with db.database.atomic():
+        box = Box.get(Box.label_identifier == box_label_identifier)
+        distribution_event = DistributionEvent.get_by_id(distribution_event_id)
```
Comment (pylipp) [inline back/boxtribute_server/models/crud.py:127]: Same line as 129 :) also you could move the box query to line 134 because in the case of an error it's superfluous

File: back/boxtribute_server/models/crud.py (hunk 11 above)
Comment (spaudanjo) [reply back/boxtribute_server/models/crud.py:187]: replace :)

File: back/boxtribute_server/models/crud.py (hunk 12 above)
Comment (spaudanjo) [reply back/boxtribute_server/models/crud.py:127]: removed the duplicated line :D 
but regarding moving it to after the error checking: the distribution_event is needed for identifying whether there is an error (it's checking the state of the event).

Comment (pylipp) [reply back/boxtribute_server/models/crud.py:127]: I meant the line `box = Box.get(...)` can be moved because it's not require for the state validation

**Review by pylipp — APPROVED** [review]

**Comment by pylipp** [comment]
[2.0 Extend new BE for Mobile Distribution Prototype](https://trello.com/c/iSwBNksN/932-20-extend-new-be-for-mobile-distribution-prototype)
